### PR TITLE
Rename outer products to tensor products

### PR DESCRIPTION
**Pull request by dham**
Basically consequential changes for:
https://bitbucket.org/mapdes/fiat/pull-requests/45/
firedrakeproject/tsfc#5
https://bitbucket.org/mapdes/ufl/pull-requests/31/

The travis branches need to backed out for merge and they all need to land together.

### CODE REVIEW

**Comment by wence-** [comment]
This looks fine I think.  There are a few `OPE` and `OPC` references I found which I will fix.  I'm inclined to squash all the changes into a single commit, which will hopefully make people who are living on branches jobs a bit easier.

**Comment by dham** [comment]
OK squash away

File: tests/extrusion/test_steady_advection_3D_extr.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def DG1(mesh):
 
 @pytest.fixture
 def W(mesh):
-    if mesh.ufl_cell()._A.cellname() == "quadrilateral":
+    if mesh.ufl_cell()._cells[0].cellname() == "quadrilateral":
```
Comment (wence-) [inline tests/extrusion/test_steady_advection_3D_extr.py]: Public API is to use `cell.sub_cells()`

**Comment by miklos1** [comment]
Can we merge this? Is 463b4e1 to be kept for the merge, or was it only for testing?

**Pull request merged**